docs/proposals/06-immutable-etcd-backups.md

```diff
@@ -0,0 +1,315 @@
+
+  - **Risk:** Failure to take a full snapshot before hibernation could delay the hibernation process.
+  - **Mitigation:** Implement robust error handling and retries. Notify operators of failures to take corrective action.
+
+- **Failed Operations:**
+
+  - **Risk:** Errors during copy or snapshot operations could lead to incomplete backups.
+  - **Mitigation:** Implement robust error handling and retries in the copier and compaction job logic. Ensure proper logging and alerting.
+
+## Operational Considerations
+
+Operators need to:
+
+- **Configure Buckets:**
+
+  - Set up buckets with appropriate immutability settings before deploying ETCD clusters.
+  - Ensure immutability periods align with organizational policies.
+
+- **Monitor Hibernation Processes:**
+
+  - Keep track of hibernated clusters and ensure that full snapshots are taken before hibernation.
+  - Verify that `EtcdCopyBackupsTask` resources are created and executed as expected.
+
+- **Review Retention Policies:**
+
+  - Set `maxBackups` and `maxBackupAge` in the `EtcdCopyBackupsTask` to manage storage utilization effectively.
+  - Configure the `--hibernation-snapshot-interval` for the compaction job if using Approach 1.
+
+## Alternatives
+
+### Object-Level Immutability Policies vs. Bucket-Level Immutability Policies
+
+An alternative to implementing immutability via bucket-level immutability policies is to use object-level immutability policies. Object-level immutability allows for more granular control over the immutability periods of individual objects within a bucket, whereas bucket-level immutability applies a uniform immutability period to all objects in the bucket.
+
+#### Feasibility Study: Immutable Backups on Cloud Providers
+
+Major cloud storage providers such as Google Cloud Storage (GCS), Amazon S3, and Azure Blob Storage (ABS) support both bucket-level and object-level immutability mechanisms to enforce data immutability.
+
+1. **Bucket-Level Immutability Policies:**
+
+   - **Applies Uniformly:** Applies a uniform immutability period to all objects within a bucket.
+   - **Immutable Objects:** Once set, objects cannot be modified or deleted until the immutability period expires.
+   - **Simplified Management:** Simplifies management by applying the same policy to all objects.
+
+2. **Object-Level Immutability Policies:**
+
+   - **Granular Control:** Allows setting immutability periods on a per-object basis.
+   - **Flexible Immutability Durations:** Offers granular control, enabling different immutability durations for individual objects.
+   - **Varying Requirements:** Can accommodate varying immutability requirements for different types of backups.
+
+#### Considerations for Object-Level Immutability
+
+Using object-level immutability provides flexibility in scenarios where certain backups require different immutability periods. However, current limitations and complexities make it less practical for immediate implementation.
```

Comment:
with object-level immutability all you need is to extend the immutability for the latest full snapshot. You do not really need to copy the full snapshot which is currently a hacky solution put in place because we are not going directly to object-level immutability policy.